src/main/kotlin/io/emeraldpay/dshackle/upstream/ethereum/BasicEthUpstreamRpcMethodsDetector.kt

```diff
@@ -0,0 +1,45 @@
+package io.emeraldpay.dshackle.upstream.ethereum
+
+import com.fasterxml.jackson.core.type.TypeReference
+import io.emeraldpay.dshackle.Global
+import io.emeraldpay.dshackle.upstream.ChainRequest
+import io.emeraldpay.dshackle.upstream.ChainResponse
+import io.emeraldpay.dshackle.upstream.Upstream
+import io.emeraldpay.dshackle.upstream.UpstreamRpcMethodsDetector
+import io.emeraldpay.dshackle.upstream.calls.DefaultEthereumMethods
+import io.emeraldpay.dshackle.upstream.rpcclient.CallParams
+import io.emeraldpay.dshackle.upstream.rpcclient.ListParams
+import reactor.core.publisher.Mono
+
+class BasicEthUpstreamRpcMethodsDetector(
+    private val upstream: Upstream,
+) : UpstreamRpcMethodsDetector(upstream) {
+    override fun detectByMagicMethod(): Mono<Map<String, Boolean>> =
+        upstream
+            .getIngressReader()
+            .read(ChainRequest("rpc_modules", ListParams()))
+            .flatMap(ChainResponse::requireResult)
+            .map(::parseRpcModules)
+            .onErrorResume {
+                log.warn("Can't detect rpc_modules of upstream ${upstream.getId()}, reason - {}", it.message)
+                Mono.empty()
+            }
+
+    override fun rpcMethods(): Set<Pair<String, CallParams>> =
+        setOf(
+            "eth_getBlockReceipts" to ListParams("latest"),
+        )
+
+    private fun parseRpcModules(data: ByteArray): Map<String, Boolean> {
+        val modules = Global.objectMapper.readValue(data, object : TypeReference<HashMap<String, String>>() {})
+        val allDisabledMethods =
+            DefaultEthereumMethods(upstream.getChain())
+                .getSupportedMethods()
+                .filter { method ->
+                    modules.all { (module, _) -> method.startsWith(module).not() }
```

Comment:
Not sure that I understand this code, could you explain pls?

Reply:
We iter over all supported methods and for each method check it start with module name (came from rpc_modules like "eth"/"web3" & etc) and if we can't find any prefix from modules for this method - we flag it as disabled